### ISSUE TITLE
OOD detection inference without access to label?

### ISSUE THREAD

**Issue by glhr** [BierOne/ood_coverage#5]
Hi,
I'm trying to adapt the implementation so that it can return the prediction and confidence for a single batch. However, it seems that the label `y` is necessary for OOD detection. Can you clarify? How can the inference be adapted so that it only takes an image as input at test-time?

Edit: seems the correct flag is not used, so this seems to work:
```python
    def postprocess(self, model: nn.Module, data: Any, spatial_func=None, progress=True, **kwargs):
        if spatial_func is None:
            spatial_func = lambda s, m: s
        transform = lambda s, m: spatial_func(s, m).detach() if len(s.shape) > 2 else s.detach()
        state_funcs = {ln: get_state_func(transform=transform, **self.hyper_dict[ln]) for ln in self.layer_names}

        with InstrumentedModel(model) as instr:
            instr.retain_layers(self.layer_names, detach=False)
            x = data.to(self.device)
            b_size = x.shape[0]
            p = model(x)
            #correct_num, b_flags = acc(p, y)
            #correct_flags = torch.ones_like(y).bool()  # we calculate cov for all samples
            b_layer_state = {}
            for j, ln in enumerate(self.layer_names):
                # we use avg-pool for last layer except the vit.encoder.ln (which employs the first clss token)
                pooling_method = "avg" if ln != "encoder.ln" else "first"
                retain_graph = False if j == len(self.layer_names) - 1 else True
                b_state = instr.retained_layer(ln)
                b_kl_grad = kl_grad(b_state, p, retain_graph=retain_graph)
                b_state, b_kl_grad = transform(b_state, pooling_method), transform(b_kl_grad, pooling_method)
                # if i == 0:
                #     print(b_kl_grad.shape, b_kl_grad.mean([2,3]).norm(dim=1,)[:3])
                out = state_funcs[ln](b_state, None, b_kl_grad)
                b_layer_state[ln] = out
            conf = self.single_ood_test(b_layer_state, b_size, **kwargs)
            pred = p.detach().argmax(1)
        
        return pred, conf
```

**Comment by BierOne** [comment]
Hi, the below is a core part of our method. It does not need the label y for the confidence calculation. You can just adapt it to your implementation.
 
```python
    b_layer_state = {}
    for j, ln in enumerate(self.layer_names):
        # we use avg-pool for last layer except the vit.encoder.ln (which employs the first clss token)
        pooling_method = "avg" if ln != "encoder.ln" else "first"
        retain_graph = False if j == len(self.layer_names) - 1 else True
        b_state = instr.retained_layer(ln)
        b_kl_grad = kl_grad(b_state, p, retain_graph=retain_graph)
        b_state, b_kl_grad = transform(b_state, pooling_method), transform(b_kl_grad, pooling_method)
        # if i == 0:
        #     print(b_kl_grad.shape, b_kl_grad.mean([2,3]).norm(dim=1,)[:3])
        out = state_funcs[ln](b_state, None, b_kl
```
